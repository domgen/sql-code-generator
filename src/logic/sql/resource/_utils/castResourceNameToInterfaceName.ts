import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
