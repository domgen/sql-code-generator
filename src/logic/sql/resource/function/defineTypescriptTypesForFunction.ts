import { TypeDefinitionOfResourceFunction } from '../../../../model/valueObjects/TypeDefinitionOfResourceFunction';
import { castResourceNameToTypescriptTypeName } from '../common/castResourceNameToTypescriptTypeName';
import { ResourceType } from '../../../../model';

export const defineTypescriptTypesForFunction = ({ definition }: { definition: TypeDefinitionOfResourceFunction }) => {
  // 0. define the typescript type namespace for this function
  const typescriptTypeName = castResourceNameToTypescriptTypeName({
    name: definition.name,
    resourceType: ResourceType.FUNCTION,
  });

  // 1. define the interface for the input
  const typescriptInterfaceInputVariablesDefinitions = definition.inputs.map((input) => {
    return `${input.name}: ${input.type.join(' | ')};`;
  });
  const inputInterfaceTypescript = `
export interface ${typescriptTypeName}Input {
  ${typescriptInterfaceInputVariablesDefinitions.join('\n  ')}
}
  `.trim();

  // 2. define the type for the output
  const outputTypeTypescript = `
export type ${typescriptTypeName}Output = ${definition.output};
  `.trim();

  // 3. return the combined typescript code
  return `
${inputInterfaceTypescript}
${outputTypeTypescript}
  `.trim();
};
