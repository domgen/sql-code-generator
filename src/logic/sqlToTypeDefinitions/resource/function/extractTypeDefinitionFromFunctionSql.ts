import { extractInputsFromFunctionSql } from './extractInputsFromFunctionSql/extractInputsFromFunctionSql';
import { extractOutputFromFunctionSql } from './extractOutputFromFunctionSql/extractOutputFromFunctionSql';
import { TypeDefinitionOfResourceFunction } from '../../../../domain/objects/TypeDefinitionOfResourceFunction';

export const extractTypeDefinitionFromFunctionSql = ({
  name,
  sql,
}: {
  name: string;
  sql: string;
}) => {
  // 1. get the input definition
  const inputs = extractInputsFromFunctionSql({ sql });

  // 2. get the output definition
  const output = extractOutputFromFunctionSql({ sql });

  // 3. return the full type
  return new TypeDefinitionOfResourceFunction({
    name,
    inputs,
    output,
  });
};
