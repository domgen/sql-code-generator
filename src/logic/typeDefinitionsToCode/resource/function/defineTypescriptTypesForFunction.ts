import { TypeDefinitionOfResourceFunction } from '../../../../model/valueObjects/TypeDefinitionOfResourceFunction';
import { castResourceNameToTypescriptTypeName } from '../../common/castResourceNameToTypescriptTypeName';
import { ResourceType } from '../../../../model';

export const defineTypescriptTypesForFunction = ({ definition }: { definition: TypeDefinitionOfResourceFunction }) => {
  // 0. define the typescript type namespace for this function
  const typescriptTypeName = castResourceNameToTypescriptTypeName({
    name: definition.name,
    resourceType: ResourceType.FUNCTION,
  });

  // 1. define the interface for the input; note: because order is the real key for the function inputs, the input is keyed by index
  const typescriptInterfaceInputVariablesDefinitions = definition.inputs.map((input, index) => {
    return `${index}: ${input.type.join(' | ')};`;
  });
  const inputInterfaceTypescript = `
export interface ${typescriptTypeName}Input {
  ${typescriptInterfaceInputVariablesDefinitions.join('\n  ')}
}
  `.trim();

  // 2. define a map of function input name => input index; since we know the name that the function gives to each input variable by index, expose that to users; maybe it will help them with debugging
  const typescriptInterfaceInputVariableIndexToNameMapDefinitions = definition.inputs.map((input, index) => {
    return `${input.name}: ${typescriptTypeName}Input['${index}'];`;
  });
  const inputByNameInterfaceTypescript = `
export interface ${typescriptTypeName}InputByName {
  ${typescriptInterfaceInputVariableIndexToNameMapDefinitions.join('\n  ')}
}
  `.trim();

  // 3. define the type for the output
  const outputTypeTypescript = `
export type ${typescriptTypeName}Output = ${definition.output};
  `.trim();

  // 4. return the combined typescript code
  return `
${inputInterfaceTypescript}
${inputByNameInterfaceTypescript}
${outputTypeTypescript}
  `.trim();
};
