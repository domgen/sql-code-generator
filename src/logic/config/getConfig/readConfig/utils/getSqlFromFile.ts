import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
