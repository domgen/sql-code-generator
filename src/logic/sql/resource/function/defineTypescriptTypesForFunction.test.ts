import { defineTypescriptTypesForFunction } from './defineTypescriptTypesForFunction';
import { extractTypeDefinitionFromFunctionSql } from './extractTypeDefinitionFromFunctionSql';
import { getSqlFromFile } from '../../../config/_utils/getSqlFromFile';

describe('defineTypescriptTypesForFunction', () => {
  it('should generate accurate looking types for this function definition', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'upsert_image',
      sql: await getSqlFromFile({ filePath: `${__dirname}/__test_assets__/upsert_image.sql` }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate accurate looking types for this other function definition', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'upsert_suggestion',
      sql: await getSqlFromFile({ filePath: `${__dirname}/__test_assets__/upsert_suggestion.sql` }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate accurate looking types for this other other function definition', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'hash_string',
      sql: await getSqlFromFile({ filePath: `${__dirname}/__test_assets__/hash_string.sql` }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
});
