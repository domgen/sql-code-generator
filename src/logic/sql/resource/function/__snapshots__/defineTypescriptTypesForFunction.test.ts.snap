// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this function definition 1`] = `
"export interface SqlFunctionUpsertImageInput {
  in_url: string | null;
  in_caption: string | null;
  in_credit: string | null;
  in_alt_text: string | null;
}
export type SqlFunctionUpsertImageOutput = number;"
`;

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this other function definition 1`] = `
"export interface SqlFunctionUpsertSuggestionInput {
  in_suggestion_source: string | null;
  in_external_id: string | null;
  in_suggested_idea_id: number | null;
  in_status: string | null;
  in_result: string | null;
  in_resultant_curated_idea_id: number | null;
}
export type SqlFunctionUpsertSuggestionOutput = number;"
`;

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this other other function definition 1`] = `
"export interface SqlFunctionHashStringInput {
  in_array: string | null;
  in_delimiter: string | null;
  in_index: number | null;
}
export type SqlFunctionHashStringOutput = string;"
`;
